### ISSUE TITLE
Automatically fill blocks with txs with different gas prices

### ISSUE THREAD

**Issue by yondonfu** [livepeer/test-harness#70]
## Problem Background

One of the most important aspects of the current PM ticket based payment workflow between Bs and Os is that Os should dynamically adjust their ticket params and price based upon the current expected gas price to include a transaction in the blockchain. In order to test this behavior, we need to be able to simulate varying gas prices required to include a transaction in the blockchain.

At the moment, blockchain enabled test network deployments typically rely on a single geth node running a private network. This geth node will accept transactions with any gas price. The Livepeer node by default use the `eth_gasPrice` RPC method to query for a suggested gas price when submitting a transaction. Under the hood, this RPC method will rely on a gas price oracle implemented on the geth node that estimates the expected gas price required to include a transaction in the next block based on the gas price of transactions included in previous blocks. Since the only transaction senders in these test network deployments are Livepeer nodes, if all nodes use the `eth_gasPrice` RPC method then all transactions included on-chain will use the same gas price resulting in the geth node's gas price oracle always returning the same gas price.

## Suggested Solution

We can simulate gas price changes by filling blocks with transactions with different gas prices. While the Livepeer nodes will be relying on the geth node's gas price oracle, an external account can send transactions with different gas prices that will change the estimated gas price returned by the gas price oracle. The transaction submission can be automated by a daemon that randomizes the gas price to include in transactions. While the daemon is running, the gas price of transactions included on-chain will be changing at regular intervals.

A prototype of this transaction filler daemon can be found [here](https://github.com/yondonfu/eth-tx-filler). The next steps for integrating this transaction filler daemon into test network deployments are:

- Dockerize the transaction filler daemon and allow configuration options to be passed in when running the container
- Generate a new account for the transaction filler in the test-harness
- Fund the new account for the transaction filler in the test-harness. The transactions sent are vanilla transfers of 0 ETH, but still funds are still required to cover the transaction fee (fairly small since a vanilla transfer is just 21000 gas)
- Add the transaction filler as a service within the test-harness (I think it can be run as a part of the `docker-compose` file similar to other existing services?)
- Automatically start the transaction filler after the geth node is started when running the test-harness in on-chain mode

**Comment by kyriediculous** [comment]
In my opinion we should be running multiple nodes (eg. 3) with different acceptable gas price settings to better simulate a real world context where miners would set their own acceptable gasprice for transactions on their nodes. 

In future iterations we can further extend upon that in the future by changing gasPrices on random intervals to further simulate a network with many nodes with different gasPrice settings. This because running more than 15 PoA nodes is sort of unfeasible (they would ping-pong too much to determine who can seal the next block) and because running that many nodes locally in a docker container would be too much for most systems to handle probably.

**Comment by yondonfu** [comment]
> In my opinion we should be running multiple nodes (eg. 3) with different acceptable gas price settings to better simulate a real world context where miners would set their own acceptable gasprice for transactions on their nodes.

I agree that running multiple nodes (multiple miners as well as multiple full nodes) would result in a more realistic network setting. However, I'm interested in taking a simpler approach with less infrastructure setup that still allows us to complete a meaningful round of payment workflow testing as a first step. Then, in subsequent rounds of testing we can add additional infrastructure that results in more and more realistic network settings. 

To clarify, I am *not* suggesting that we modify a miner's minimum gas price (i.e. the minimum gas price to enter the miner's tx pool) right now, but rather that we should have the miner accept any gas price and we can fill blocks with transactions with different gas prices so that when O queries for a suggested gas price it will receive a gas price that reflects the gas prices used in the most recent blocks (which is how all gas price estimators including geth's gas price oracle and the ethgasstation API work). This setup alone should allow us to answer the one of the most important questions of this round of testing: "can O reliably dynamically adjust its ticket params and price in the face of fluctuating gas prices?". This setup should be possible with a single miner (technically a PoA sealer).

I see the following as potential next steps to consider for later rounds of testing:

- Increase the block time within the private network (it is 1 second right now)
- Vary the tx submission frequency for the tx filler daemon and include heavier txs so that we can better simulate alternating between a congested vs. uncongested network. A congested network combined with longer than 1 second block times should also automatically lead to slower tx confirmation times which will force us to make improvements to the client that robustly handle these scenarios (i.e. if a ticket redemption tx is pending for a long time, automatically bumping its gas price)
- Add additional PoA sealers to the network
- Run a cluster of full nodes for the Livepeer nodes to talk to instead of the Livepeer nodes directly talking to the single miner on the network

**Comment by kyriediculous** [comment]
> To clarify, I am not suggesting that we modify a miner's minimum gas price

I know and that's totally fair. 

But one interesting test case I find is that a transaction for ticket redemption is included within a reasonable amount of future blocks. 

Allow me to illustrate with the most extreme case. 
1. O submits a ticket redemption, which would be redeemable according to available max-float - sits in mempool for over 7 days
2. B creates an unlock transaction - included in the next block
3. B creates a withdrawal transaction after unlock is processed - included in teh next block
4. O's ticket redemption ticket finally gets picked up and mined 
5. O is sad because the redemption failed simply due to race conditions. 

For this case having gas price races between tx's to get included into the next block is one thing, simulating different nodes with different gas prices is an other as different nodes will have different gas price requirements for a tx to be included. But perhaps this is not as much related to this issue. 

Or is this what is meant by "is redeemed immeadiatly" in the acceptance tests? I interpreted that as a transaction is submitted immeadiatly, but that transaction is not necessarily included in the next block immeadiatly.

**Comment by yondonfu** [comment]
> Allow me to illustrate with the most extreme case.

Yeah this would be the expected node behavior right now. We'll need to add automatic redemption tx replacements for txs that sit pending for a long time (since it is possible that the gas price estimate used for the first tx is not sufficient). livepeer/go-livepeer#654 hints at this feature, but the issue description could be updated to highlight tx replacements. Since, we know that the node isn't capable of handling this scenario right now, I think we can leave this scenario for a subsequent round of testing (when we've implemented this feature so we expect the node to be able to handle this scenario).

> For this case having gas price races between tx's to get included into the next block is one thing, simulating different nodes with different gas prices is an other as different nodes will have different gas price requirements for a tx to be included.

From the PoV of a client that is submitting a tx, these two cases should look the same - all the client sees is its tx not being included in mined blocks. The reason for the tx not being included in mined blocks could be that the avg gas price of other txs is much higher so this tx is low priority or it could be that the miners minimum tx pool gas price is higher than the gas price of this tx. Either way, the client doesn't know the actual reason. We can certainly configure different minimum gas prices on miners in a future round of testing to get closer to a real network setting, but for this round of testing this configuration doesn't help us test any behavior that we can't already test just by filling blocks with txs with different gas prices.

> I interpreted that as a transaction is submitted immeadiatly, but that transaction is not necessarily included in the next block immeadiatly.

I did mean that the transaction is submitted immediately, but I'm also assuming timely tx inclusion on-chain for this round of testing and leaving long tx confirmation times as a concern for a subsequent round of testing. Added a note in the spreadsheet for clarification.
